core/src/main/scala/algebra/ring/Signed.scala

```diff
@@ -0,0 +1,141 @@
+ */
+trait Signed[@sp(Byte, Short, Int, Long, Float, Double) A] extends Any with Order[A] {
+
+  /**
+   * Returns Zero if `a` is 0, Positive if `a` is positive, and Negative is `a` is negative.
+   */
+  def sign(a: A): Signed.Sign = Signed.Sign(signum(a))
+
+  /**
+   * Returns 0 if `a` is 0, 1 if `a` is positive, and -1 is `a` is negative.
+   */
+  def signum(a: A): Int
+
+  /**
+   * An idempotent function that ensures an object has a non-negative sign.
+   */
+  def abs(a: A): A
+
+  def isSignZero(a: A): Boolean = signum(a) == 0
+  def isSignPositive(a: A): Boolean = signum(a) > 0
+  def isSignNegative(a: A): Boolean = signum(a) < 0
+
+  def isSignNonZero(a: A): Boolean = signum(a) != 0
+  def isSignNonPositive(a: A): Boolean = signum(a) <= 0
+  def isSignNonNegative(a: A): Boolean = signum(a) >= 0
+}
+
+trait SignedFunctions[S[T] <: Signed[T]] extends cats.kernel.OrderFunctions[S] {
+  def sign[@sp(Int, Long, Float, Double) A](a: A)(implicit ev: S[A]): Signed.Sign =
+    ev.sign(a)
+  def signum[@sp(Int, Long, Float, Double) A](a: A)(implicit ev: S[A]): Int =
+    ev.signum(a)
+  def abs[@sp(Int, Long, Float, Double) A](a: A)(implicit ev: S[A]): A =
+    ev.abs(a)
+  def isSignZero[@sp(Int, Long, Float, Double) A](a: A)(implicit ev: S[A]): Boolean =
+    ev.isSignZero(a)
+  def isSignPositive[@sp(Int, Long, Float, Double) A](a: A)(implicit ev: S[A]): Boolean =
+    ev.isSignPositive(a)
+  def isSignNegative[@sp(Int, Long, Float, Double) A](a: A)(implicit ev: S[A]): Boolean =
+    ev.isSignNegative(a)
+  def isSignNonZero[@sp(Int, Long, Float, Double) A](a: A)(implicit ev: S[A]): Boolean =
+    ev.isSignNonZero(a)
+  def isSignNonPositive[@sp(Int, Long, Float, Double) A](a: A)(implicit ev: S[A]): Boolean =
+    ev.isSignNonPositive(a)
+  def isSignNonNegative[@sp(Int, Long, Float, Double) A](a: A)(implicit ev: S[A]): Boolean =
+    ev.isSignNonNegative(a)
+}
+
+object Signed extends SignedFunctions[Signed] {
+
+  /** Signed implementation for additive commutative monoids */
+  trait forAdditiveCommutativeMonoid[A] extends Any with Signed[A] with AdditiveCommutativeMonoid[A] {
+    def signum(a: A): Int = {
+      val c = compare(a, zero)
+      if (c < 0) -1
+      else if (c > 0) 1
+      else 0
+    }
+  }
+
+  /** Signed implementation for additive commutative groups */
+  trait forAdditiveCommutativeGroup[A] extends Any with forAdditiveCommutativeMonoid[A] with AdditiveCommutativeGroup[A] {
+    def abs(a: A): A = if (compare(a, zero) < 0) negate(a) else a
+  }
+
+  def apply[A](implicit s: Signed[A]): Signed[A] = s
+
+  /**
+   * A simple ADT representing the `Sign` of an object.
+   */
+  sealed abstract class Sign(val toInt: Int) {
+    def unary_- : Sign = this match {
+      case Positive => Negative
+      case Negative => Positive
+      case Zero     => Zero
+    }
+
+    def *(that: Sign): Sign = Sign(this.toInt * that.toInt)
+
+    def **(that: Int): Sign = this match {
```

Comment:
Can we add a comment that this is exponentiation?

Reply:
Sure!